syspagegen/syspagegen.c

```diff
@@ -0,0 +1,822 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Tool to generate syspage based on plo scripts
+ *
+ * Copyright 2022 Phoenix Systems
+ *
+ * Author: Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <errno.h>
+#include <stdio.h>
+#include <ctype.h>
+#include <stdint.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <string.h>
+#include <fcntl.h>
+#include <getopt.h>
+#include <sys/stat.h>
+#include <sys/queue.h>
+
+#include "syspage.h"
+
+
+#define ALIGN_ADDR(addr, align) (align ? ((addr + (align - 1)) & ~(align - 1)) : addr)
+#define CAST_SYSPTR(type, ptr)  ((type)(sysgen_common.buff + ptr - (sysgen_common.pkernel + sysgen_common.offs)))
+
+/* Reserve +1 for terminating NULL pointer in conformance to C standard */
+#define SIZE_CMD_ARGV     (10 + 1)
+#define SIZE_CMD_ARG_LINE 181
+
+
+struct phfs_alias_t {
+	char name[32];
+	addr_t addr;
+	size_t size;
+
+	LIST_ENTRY(phfs_alias_t) ptrs;
+};
+
+
+typedef struct {
+	const char name[12];
+	const int (*run)(int, char *[]);
+} cmd_t;
+
+
+struct {
+	addr_t pkernel;
+	off_t offs;
+	size_t maxsz;
+
+	syspage_t *syspage;
+	uint8_t *buff;
+
+	LIST_HEAD(plist_t, phfs_alias_t) aliases;
+} sysgen_common;
+
+
+extern int sysgen_cmdAlias(int argc, char *argv[]);
+extern int sysgen_cmdMap(int argc, char *argv[]);
+extern int sysgen_cmdApp(int argc, char *argv[]);
+extern int sysgen_cmdConsole(int argc, char *argv[]);
+
+
+static const cmd_t cmds[] = {
+	{ .name = "alias", .run = sysgen_cmdAlias },
+	{ .name = "map", .run = sysgen_cmdMap },
+	{ .name = "app", .run = sysgen_cmdApp },
+	{ .name = "console", .run = sysgen_cmdConsole }
+};
+
+enum { flagSyspageExec = 0x01 };
+
+
+static sysptr_t sysgen_buffAlloc(size_t sz)
+{
+	sysptr_t ptr;
+	size_t newSz = ALIGN_ADDR(sysgen_common.syspage->size + sz, sizeof(long long));
+
+	if (newSz >= sysgen_common.maxsz) {
+		fprintf(stderr, "Cannot allocate size 0x%x; current buffer size 0x%x\n", (unsigned int)sz, sysgen_common.syspage->size);
+		return 0;
+	}
+
+	ptr = sysgen_common.syspage->size + sysgen_common.pkernel + sysgen_common.offs;
+	sysgen_common.syspage->size = newSz;
+
+	return ptr;
+}
+
+
+static struct phfs_alias_t *sysgen_aliasFind(const char *name)
+{
+	struct phfs_alias_t *alias;
+
+	LIST_FOREACH(alias, &sysgen_common.aliases, ptrs) {
+		if (strcmp(alias->name, name) == 0)
+			return alias;
+	}
+
+	return NULL;
+}
+
+
+int sysgen_cmdAlias(int argc, char *argv[])
+{
+	char *end;
+	unsigned int i;
+	addr_t addr = 0;
+	size_t size = 0, len;
+	struct phfs_alias_t *alias;
+
+	if (argc != 4) {
+		fprintf(stderr, "\n%s: Wrong argument count", argv[0]);
+		return -1;
+	}
+
+	for (i = 2; i < 4; ++i) {
+		if (i == 2)
+			addr = strtoul(argv[i], &end, 0);
+		else if (i == 3)
+			size = strtoul(argv[i], &end, 0);
+
+		if (*end) {
+			fprintf(stderr, "\n%s: Wrong arguments", argv[0]);
+			return -1;
+		}
+	}
```

Comment:
imho better to unroll this loop - too much is `i` dependent

Reply:
I have reused the working code from `plo` which is tested and works fine.